Dockerfile

```diff
@@ -1,45 +1,78 @@
-FROM ruby:2.5.3-alpine
+FROM ruby:2.5.3-alpine as base
 
 ENV APP_ROOT /usr/src/app
 WORKDIR $APP_ROOT
 
-# =============================================
-# System layer
-
-# Will invalidate cache as soon as the Gemfile changes
-COPY Gemfile Gemfile.lock $APP_ROOT/
-
-# * Setup system
-# * Install Ruby dependencies
-RUN apk add --update \
-    build-base \
+RUN apk add --no-cache \
     netcat-openbsd \
     git \
     nodejs \
     postgresql-dev \
     mysql-dev \
     tzdata \
-    curl-dev \
- && rm -rf /var/cache/apk/* \
+    curl-dev
+
+# Will invalidate cache as soon a the Gemfile changes
+COPY Gemfile Gemfile.lock $APP_ROOT/
+
+# * Setup system
+# * Install common Ruby dependencies
+RUN apk add --no-cache build-base \
  && bundle config --global frozen 1 \
- && bundle install --without test --jobs 2 \
- && gem install foreman
+ && bundle install \
+    --without development test production \
+    --jobs $(grep -c ^processor /proc/cpuinfo) \
+ && gem install foreman \
+ && apk del build-base
 
-# ========================================================
-# Application layer
+# Startup
+CMD ["bin/docker-start"]
+
+####################
+
+FROM base as development
+
+# * Install development dependencies
+RUN apk add --no-cache build-base \
+ && bundle install \
+    --with development \
+    --without test production \
+    --jobs $(grep -c ^processor /proc/cpuinfo) \
+ && apk del build-base
 
 # Copy application code
 COPY . $APP_ROOT
 
-# Precompile assets for a production environment.
-# This is done to include assets in production images on Dockerhub.
-RUN RAILS_ENV=production bundle exec rake assets:precompile
-
 # * Generate the docs
 # * Make files OpenShift conformant
 RUN RAILS_ENV=development bin/rails api_docs:generate \
- && chgrp -R 0 $APP_ROOT \
+ && chown -R 1000:0 $APP_ROOT \
  && chmod -R g=u $APP_ROOT
 
-# Startup
-CMD ["bin/docker-start"]
+USER 1000
```

Comment:
From the docs... 

> The USER instruction sets the user name (or UID) and optionally the user group (or GID) to use when running the image and for any RUN, CMD and ENTRYPOINT instructions that follow it in the Dockerfile.

Doesn't that mean this should be before all the other commands in each section?

Reply:
Unfortunately this would mean running apk with user 1000 and that won't work. Since `bundle install` commands are all surrounded by apk commands, the only place where we can set the `USER` is at the end of each target.

However, AFAIK bundler and RubyGems does not have the same security issue than, for instance, npm: there's no external code run during installation phase, so I believe there's no reason for the build environment to be compromised (from that vector). In the end, it's only there to not run Octobox as a root user, as this is considered a bad practice (see [here](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#user)).

Reply:
Rubygems does actually allow for arbitrary code execution during installation via the option to build native extensions: http://blog.costan.us/2008/11/post-install-post-update-scripts-for.html although it requires jumping through more hoops than with npm.

Reply:
Ha too bad :/ I'm not really sure what to do then. I tried to find out articles about "dockerizing" ruby apps, but almost none talks about `USER` instruction. The only one I found using this instruction, does so after installing the dependencies (see [here](https://blog.kontena.io/dockerizing-ruby-application#exampleapplication)).

Do you think this is not an issue in the end, or should I do the required changes to not run bundler as root (that would mean we couldn't install/deinstall build-tools before/after running bundler)?